matchmaker.js

```diff
@@ -0,0 +1,146 @@
+import Player from './player.js'
+import GameRoom from './gameroom.js';
+import RoomManager from './roommanager.js'
+import { WebSocketServer } from 'ws'
+import { v4 as uuidv4 } from 'uuid';
+
+import { upload_to_blob_storage } from './blobstorage.js';
+
+class Matchmaker {
+    constructor(room_manager) {
+        this.banned_characters = ["carmine"]
+        this.room_manager = room_manager
+        this.match_queues = {}
+
+        this.match_types = {
+            "untimed" : {
+              "starting_timer" : 0,
+              "enforce_timer" : false,
+              "minimum_time_per_choice" : 0},
+            "timed" : {
+              "starting_timer" : 12,
+              "enforce_timer" : true,
+              "minimum_time_per_choice" : 20},
+            "speed" : {
+              "starting_timer" : 6,
+              "enforce_timer" : true,
+              "minimum_time_per_choice" : 10}
+        }
+    
+    for (queue_type in this.match_types.keys) {
+            this.match_queues[queue_type] = {
+                "has_player_waiting" : false,
+                "waiting_player" : player = null,
+                "waiting_json_data" : object = null,
+                "waiting_version" : Number
+            }
+        }
+    }
+
+    remove_from_matchmaking(player) {
+        for (queue in this.match_queues.keys) {
+            if (queue["waiting_player"] == player) {
+                this.match_queues[queue]["has_player_waiting"] = false
+            }
+        }
+    }
+
+    join_matchmaking(ws, json_data) {
+        // Check if json_data is an object
+        if (typeof json_data !== 'object' || json_data === null) {
+            console.log("join_matchmaking json is not an object")
+            return false
+        }
+        // Check if 'deck_id' field exists in the object
+        if (!('deck_id' in json_data)) {
+            console.log("join_matchmaking does not have 'deck_id' fields")
+            return false
+        }
+        if (!(typeof json_data.deck_id === 'string' && typeof json_data.deck_id === 'string')) {
+            console.log("join_matchmaking 'deck_id' fields are not strings")
+            return false
+        }
+        if (!('version' in json_data && typeof json_data.version === 'string')) {
+            console.log("join_matchmaking does not have 'version' field")
+            return false
+        }
+        if (!('match_type' in json_data && typeof json_data.match_type === 'string')) {
+            console.log("join_matchmaking does not have 'match_type' field")
+        }
+
+        var joining_player = active_connections.get(ws)
+        if (joining_player === undefined) {
+            console.log("join_matchmaking Player is undefined")
+            return false
+        }
+
+        var player_join_version = json_data.version
+        var deck_id = json_data.deck_id
+        joining_player.set_deck_id(deck_id)
+        var desired_match_type = json_data.match_type
```

Comment:
please validate that desired_match_type is actually one of the valid queue types, if not, return an error message.

Reply:
The goal is to have as much error handling as possible so bad messages can't crash the server.